File: sl2/jsonout.py
"""JSON output: the same parsed save the Markdown writer gets, as a machine-readable
document against a published schema.

The Markdown is written for a person (or an LLM) to read; this is written for a
program. Both start from `parse_save`, so a field can never say one thing in one
format and something else in the other.

Two rules shape the document. Nothing is invented to fill a hole — a field the save
does not carry is simply absent, exactly as in the Markdown, so a consumer can tell
"not in this game" from "zero". And nothing about the machine that produced it is
guessed: the environment block holds only what the caller passed on the command line.
"""
import json
import os
from collections import OrderedDict
from datetime import datetime, timezone

from .convert import REPO_URL

## @brief Where the schema is published. Static file at the site root, so a consumer
#  can resolve it without cloning anything.
SCHEMA_URL = "https://darthdemono.github.io/sl2-analyzer/schema.json"

## @brief Schema version, semver. MINOR for a new optional field, MAJOR for anything
#  that would break a reader which trusted the previous shape.
SCHEMA_VERSION = "1.0.0"

## @brief Environment keys the schema names explicitly. Any other key is still
#  accepted and written through — this list is what gets documented and type-checked,
#  not a whitelist. The point of the block is that a save alone cannot tell you which
#  store sold the game, which patch it ran, or what it ran under.
KNOWN_META = ("source", "version", "dlc", "os", "launcher", "proton", "gamemode",
              "mangohud", "notes")


##
# @brief Normalise a CLI metadata key: lowercase, spaces and dashes to underscores.
# @details "Proton version" and "proton-version" are the same key. The value is left
#  exactly as typed — only the key is canonicalised, because the key is what a
#  consumer looks up.
# @param key Raw key as typed. @return The canonical form.
def meta_key(key):
    return key.strip().lower().replace(" ", "_").replace("-", "_")


##
# @brief Build the environment block from repeated `--meta key=value` arguments and
#        an optional JSON file.
# @details A key given more than once becomes a LIST, in the order given — that is how
#  `--meta dlc=X --meta dlc=Y` says two DLCs, with no comma-splitting guesswork (item
#  and boss names are full of commas, so splitting on one would be a bug waiting).
#  The JSON file is merged first so an explicit `--meta` on the command line wins.
# @param pairs List of "key=value" strings, or None.
# @param path  Path to a JSON object to merge underneath, or None.
# @return An OrderedDict, empty when nothing was passed.
# @throws ValueError on an argument with no "=", or a JSON file that is not an object.
def parse_meta(pairs, path=None):
    meta = OrderedDict()
    if path:
        with open(path, encoding="utf-8") as f:
            loaded = json.load(f)
        if not isinstance(loaded, dict):
            raise ValueError(f"{path}: expected a JSON object at the top level")
        for k, v in loaded.items():
            meta[meta_key(k)] = v
    for pair in pairs or []:
        if "=" not in pair:
            raise ValueError(f"--meta expects key=value, got {pair!r}")
        key, value = pair.split("=", 1)
        key, value = meta_key(key), value.strip()
        if key in meta:
            # Repeat means "and also", so the first repeat turns the value into a list.
            meta[key] = (meta[key] if isinstance(meta[key], list) else [meta[key]]) + [value]
        else:
            meta[key] = value
    return meta


##
# @brief Make a parsed value safe for json.dump, without changing what it says.
# @details The parser uses sets for boss evidence and tuples for fixed-shape rows;
#  both become lists, and a set is sorted so two runs of the same save produce the
#  same bytes. Everything else passes through untouched.
def jsonable(value):
    if isinstance(value, dict):
        return OrderedDict((k, jsonable(v)) for k, v in value.items())
    if isinstance(value, set):
        return sorted(value)
    if isinstance(value, (list, tuple)):
        return [jsonable(v) for v in value]
    return value


##
# @brief One character as a JSON object: its slot number, then every field the parse
#        actually set.
# @details Absent stays absent. The keys are the parser's own, which is deliberate —
#  the Markdown, the web app and this document all name a field the same thing.
# @param slot_no 1-based slot number. @param ch The character dict.
def character_json(slot_no, ch):
    out = OrderedDict([("slot", slot_no)])
    for key, value in ch.items():
        if value is None or value == [] or value == {}:
            continue
        out[key] = jsonable(value)
    return out


##
# @brief Build the whole JSON document for a parsed save.
# @param save     A @ref sl2.convert.SaveData.
# @param filename The source filename, recorded so an export can be traced back.
# @param meta     The environment block from @ref parse_meta, or None.
# @return A dict ready for json.dump.
def build_json(save, filename, meta=None):
    cfg = save.cfg
    source = OrderedDict([
        ("filename", os.path.basename(filename)),
        ("game", save.game),
        ("game_title", cfg["title"]),
        ("support_tier", cfg["tier"]),
    ])
    # Both are properties of the FILE, not of any character, and both are frequently
    # absent — DS2 has no version word, and only ER carries a regulation version.
    if save.version is not None:
        source["save_format_version"] = save.version
    if save.patch is not None:
        source["game_patch"] = save.patch

    doc = OrderedDict([
        ("$schema", SCHEMA_URL),
        ("schema_version", SCHEMA_VERSION),
        ("generated", datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")),
        ("tool", OrderedDict([("name", "sl2-analyzer"), ("url", REPO_URL)])),
        ("source", source),
    ])
    if meta:
        doc["environment"] = OrderedDict(meta)
    doc["characters"] = [character_json(i - cfg["slots"].start + 1, ch)
                         for i, ch in save.characters]
    return doc

File: app/jsonout.js
// The browser's JSON export: the same document the Python CLI writes with -o out.json,
// against the same published schema. Held byte-for-byte by scratch/json_harness.mjs, the
// way markdown.js is held by md_harness.mjs.
//
// Faithful to sl2/jsonout.py. If you change a key, a value shape, or the order they are
// written in, change it there too — a consumer validating against schema.json should not
// be able to tell which front end produced the file.

import { REPO_URL } from "./markdown.js";

/** Where the schema is published. Static file at the site root. */
export const SCHEMA_URL = "https://darthdemono.github.io/sl2-analyzer/schema.json";

/** Semver for the document format. MINOR adds an optional field, MAJOR breaks a reader. */
export const SCHEMA_VERSION = "1.0.0";

/**
 * ISO-8601 UTC to the second, which is what the Python writes. `toISOString` includes
 * milliseconds and those would be the only difference between the two exports.
 */
function stampUtc(now = new Date()) {
  return now.toISOString().replace(/\.\d{3}Z$/, "Z");
}

/**
 * True for the values the Python skips: None, an empty list, an empty dict. Everything
 * else is written, including 0 and false, which are real readings rather than gaps.
 * @param {*} v
 */
function isEmpty(v) {
  if (v === null || v === undefined) return true;
  if (Array.isArray(v)) return v.length === 0;
  if (v instanceof Map || v instanceof Set) return v.size === 0;
  return typeof v === "object" && Object.keys(v).length === 0;
}

/**
 * Make a parsed value safe for JSON.stringify without changing what it says. The parser
 * hands back plain arrays and objects already; a Map or Set can only appear if a future
 * reader introduces one, and this keeps that from silently serialising as `{}`.
 * @param {*} v
 */
function jsonable(v) {
  if (v instanceof Map) return Object.fromEntries([...v].map(([k, x]) => [k, jsonable(x)]));
  if (v instanceof Set) return [...v].sort();
  if (Array.isArray(v)) return v.map(jsonable);
  if (v && typeof v === "object") {
    const out = {};
    for (const [k, x] of Object.entries(v)) out[k] = jsonable(x);
    return out;
  }
  return v;
}

/**
 * One character: its slot number, then every field the parse actually set. Absent stays
 * absent — that is the contract the schema documents, so a consumer can tell "this game
 * does not store it" from "it is zero".
 * @param {number} slotNo 1-based slot number
 * @param {object} ch the character object
 */
function characterJson(slotNo, ch) {
  const out = { slot: slotNo };
  for (const [key, value] of Object.entries(ch)) {
    if (isEmpty(value)) continue;
    out[key] = jsonable(value);
  }
  return out;
}

/**
 * Build the whole JSON document for a parsed save.
 * @param {object} result what parseSave returned
 * @param {string} filename the source filename, recorded so an export can be traced back
 * @param {object} [meta] caller-supplied environment, or null. The page has no way to know
 *        any of it, so it is only ever present when something passed it in.
 * @returns {object} ready for JSON.stringify
 */
export function buildJson(result, filename, meta = null) {
  const source = {
    filename: (filename || "").split(/[/\\]/).pop(),
    game: result.game,
    game_title: result.title,
    support_tier: result.tier || "full",
  };
  // Both are properties of the FILE, not of any character, and both are often absent —
  // DS2 has no version word and only Elden Ring carries a regulation version.
  if (result.saveVersion != null) source.save_format_version = result.saveVersion;
  if (result.gamePatch != null) source.game_patch = result.gamePatch;

  const doc = {
    $schema: SCHEMA_URL,
    schema_version: SCHEMA_VERSION,
    generated: stampUtc(),
    tool: { name: "sl2-analyzer", url: REPO_URL },
    source,
  };
  if (meta && Object.keys(meta).length) doc.environment = { ...meta };
  doc.characters = result.characters.map(({ slot, ch }) => characterJson(slot, ch));
  return doc;
}

/** The document as the text that gets written to a file — 2-space indent, like the CLI. */
export function buildJsonText(result, filename, meta = null) {
  return `${JSON.stringify(buildJson(result, filename, meta), null, 2)}\n`;
}
